### ISSUE TITLE
Cannot construct vector with complex coordinates (StackOverflowError)

### ISSUE THREAD

**Issue by Batmaev** [JuliaGeometry/Meshes.jl#834]
I cannot construct a `Vector` with `Complex` or `Dual` coordinates.

Example:
```julia
Vec(0)                   # works
Vec(0 + 0im)             # StackOverflowError
Vec(ForwardDiff.Dual(0)) # StackOverflowError
```
Previously it was possible to use these coordinates. I think the issue was introduced in PR #763. This PR restricted vector coordinates to `AbstractFloat`, while
```julia
isa(0 + 0im,             AbstractFloat) # false
isa(ForwardDiff.Dual(0), AbstractFloat) # false
```

**Comment by juliohm** [comment]
We are currently working on a major refactor of the code to use CoordRefSystems.jl instead: https://github.com/JuliaEarth/CoordRefSystems.jl

After the refactor is finished, we can look into the use cases and try to decide the level of flexibility supported.

**Comment by juliohm** [comment]
In the meantime, can you please elaborate on the two use cases? How are you using Complex and Dual?

Perhaps Meshes.jl`s scope doesn't include your use case with Complex coordinates. We are aiming applications with geometries in the physical world with coordinates in R^n.

**Comment by Batmaev** [comment]
I use Dual numbers for automatic differentiation. I mentioned Complex numbers only because with them the issue is simpler to reproduce, since they are included in the standard library.

-------

I am modeling light curves of symbiotic binary stars. The surface of the first star is distorted by the gravity of the second one, and vice versa. The level of distortion is determined by the mass ratio of the two components.

I use Meshes.jl to construct the surface of the star. Then, I integrate over this surface to calculate the visible light intensity.

My goal is to determine the mass ratio based on available measurements. As part of this process, I calculate the derivative of the total light intensity with respect to the mass ratio, using Dual numbers.

The differentiation occurs within the Turing.jl library and Markov chain samplers.

-------

I believe the spirit of Julia is to allow users to employ different types and take advantage of multiple dispatch, not limiting ourselves to just Float64.

**Comment by juliohm** [comment]
> I use Dual numbers for automatic differentiation. I mentioned Complex numbers only because with them the issue is simpler to reproduce, since they are included in the standard library.

Makes total sense. We have autodiff in mind and would like to handle it in the CoordRefSystems.jl side.

> I am modeling light curves of symbiotic binary stars. The surface of the first star is distorted by the gravity of the second one, and vice versa. The level of distortion is determined by the mass ratio of the two components.
> 
> I use Meshes.jl to construct the surface of the star. Then, I integrate over this surface to calculate the visible light intensity.
> 
> My goal is to determine the mass ratio based on available measurements. As part of this process, I calculate the derivative of the total light intensity with respect to the mass ratio, using Dual numbers.
> 
> The differentiation occurs within the Turing.jl library and Markov chain samplers.

Very interesting use case. Let's see how that can be done in a future release of the project with the new CRS support.

> I believe the spirit of Julia is to allow users to employ different types and take advantage of multiple dispatch, not limiting ourselves to just Float64.

Definitely. That is why we've been pushing native Julia packages all the way, to give users this flexibility. We just need to nail down the scope in terms of types as well, because machine representation is not the same thing as scientific representation. CRSs currently use floats to represent coordinates, but nothing inhibits us from using another representation in the future, if it is in the scope of the project. Complex coordinates are not in the scope for example.

BTW, we already use autodiff with Zygote.jl in CoordRefSystems.jl and everything works out of the box, using units. Are you tied to Dual or would be fine with other autodiff approaches?

**Comment by mikeingold** [comment]
> In the meantime, can you please elaborate on the two use cases? How are you using Complex and Dual?

I also ran into this error constructing complex-typed geometries earlier this year. If I remember correctly, this seemed to work earlier this year but at some point maybe two months ago it stopped working.

I didn’t have a particularly compelling use case, but I was testing out enabling contour integrals to be handled in [MeshIntegrals.jl](https://github.com/mikeingold/MeshIntegrals.jl) by constructing `Ring`/`Rope`s of `Point{1,<:Complex}`. Unfortunately, even before the support in Meshes seemed to be cut off, the integrals would run but produced wrong results. I suspect that issue was on my side because the typical Meshes features I was hitting seemed to produce correct results, despite it being a case that I suspect it was never intended to support. I had it labeled as a back-burner issue to address later since it seemed a pretty esoteric usage Ann’s was a low priority for me, and once it started throwing errors instead of simply failing tests, I made the call to just drop the feature.

On the note of automatic differentiation, I’d be very happy to be able to AD the Jacobian/gradient of parameterization functions. I currently have a method in MeshIntegrals for approximating these via finite-difference but it’s not very elegant.
